### PR TITLE
Introduce ELF loader with networking support

### PR DESCRIPTION
**Pull request by razvand**
Add `elfloader-net/` directory (with networking support)

- `Config.uk`: ELF Loader configuration
- `rootfs/`: setup for building Linux ELFs, with simple C HTTP server
- `Makefile`: for building the ELF Loader
- `Makefile.uk`: empty placeholder file required by build system
- `fc.x86_64.json`: Firecracker configuration
- `README.md`: instructions

### CODE REVIEW

**Review by StefanJum — CHANGES_REQUESTED** [review]
I get a lot of build errors related to lwip, see the comments.

Also, is there a reason why we have two separate elfloader application, one with and one without networking? Why would we not just use the one with networking enabled?
I think having 2 instances of the same app might get hard to maintain over time, but if there is a good reason for it then it's ok.

File: elfloader-net/README.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+For a quick setup, run the commands below.
+Note that you still need to install the [requirements](../README.md#requirements).
+
+To build and run a simple HTTP server Linux ELF using the Unikraft ELF Loader, use the commands below:
+
+```console
+test -d ../repos/unikraft || git clone https://github.com/unikraft/unikraft ../repos/unikraft
+test -d ../repos/libs/libelf || git clone https://github.com/unikraft/lib-libelf ../repos/libs/libelf
+test -d "../repos/libs/lwip" || git clone https://github.com/unikraft/lib-lwip ../repos/libs/lwip
```
[the remark this answers is not in the document]
Comment (razvand) [reply elfloader-net/README.md:22]: Yes, we need to also clone that, will do.

**Comment by razvand** [comment]
> Also, is there a reason why we have two separate elfloader application, one with and one without networking? Why would we not just use the one with networking enabled?

The reason is too keep things simple and quick to build for the `elfloader-basic` app, without the need to build `lwip`. Additionally, it can be a good test case to build / run without networking support. We could do some sed-ing / processing to not include LWIP, but I think it's cleaner to have them separate.

File: elfloader-net/README.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+CONFIG_APPELFLOADER_HFS_REPLACEEXIST=y
+CONFIG_LIBVFSCORE=y
+CONFIG_LIBVFSCORE_AUTOMOUNT_UP=y
+CONFIG_LIBRAMFS=y
+CONFIG_LIBUKCPIO=y
+CONFIG_LIBDEVFS=y
+CONFIG_LIBDEVFS_AUTOMOUNT=y
+CONFIG_LIBLWIP=y
+CONFIG_LIBUKNETDEV_EINFO_LIBPARAM=y'
+UK_DEFCONFIG=/tmp/defconfig make defconfig
```
[the remark this answers is not in the document]
Comment (razvand) [reply elfloader-net/README.md:47]: Are you sure you are running the commands in the `README.md`? Namely:

```console
test -d ../repos/unikraft || git clone https://github.com/unikraft/unikraft ../repos/unikraft
test -d ../repos/libs/libelf || git clone https://github.com/unikraft/lib-libelf ../repos/libs/libelf
test -d "../repos/libs/lwip" || git clone https://github.com/unikraft/lib-lwip ../repos/libs/lwip
make distclean
> /tmp/defconfig echo 'CONFIG_PLAT_KVM=y
CONFIG_KVM_VMM_QEMU=y
CONFIG_ARCH_X86_64=y
CONFIG_APPELFLOADER_DEPENDENCIES=y
CONFIG_APPELFLOADER_ARCH_PRCTL=y
CONFIG_APPELFLOADER_BRK=y
CONFIG_APPELFLOADER_CUSTOMAPPNAME=y
CONFIG_APPELFLOADER_STACK_NBPAGES=128
CONFIG_APPELFLOADER_VFSEXEC_EXECBIT=n
CONFIG_APPELFLOADER_VFSEXEC=y
CONFIG_APPELFLOADER_HFS=y
CONFIG_APPELFLOADER_HFS_ETCRESOLVCONF=y
CONFIG_APPELFLOADER_HFS_ETCHOSTS=y
CONFIG_APPELFLOADER_HFS_ETCHOSTNAME=y
CONFIG_APPELFLOADER_HFS_REPLACEEXIST=y
CONFIG_LIBVFSCORE=y
CONFIG_LIBVFSCORE_AUTOMOUNT_UP=y
CONFIG_LIBRAMFS=y
CONFIG_LIBUKCPIO=y
CONFIG_LIBDEVFS=y
CONFIG_LIBDEVFS_AUTOMOUNT=y
CONFIG_LIBLWIP=y
CONFIG_LIBUKNETDEV_EINFO_LIBPARAM=y'
UK_DEFCONFIG=/tmp/defconfig make defconfig
make -j $(nproc)
make -C rootfs/
test -f initrd.cpio || ../repos/unikraft/support/scripts/mkcpio initrd.cpio ./rootfs/
sudo qemu-system-x86_64 \
    -nographic \
    -m 64 \
    -cpu max \
    -netdev bridge,id=en0,br=virbr0 -device virtio-net-pci,netdev=en0 \
    -append "elfloader_qemu-x86_64 netdev.ip=172.44.0.2/24:172.44.0.1::: vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /c-server" \
    -kernel out/elfloader_qemu-x86_64 \
    -initrd ./initrd.cpio
```

File: elfloader-net/README.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+make -j $(nproc)
+make -C rootfs/
+test -f initrd.cpio || ../repos/unikraft/support/scripts/mkcpio initrd.cpio ./rootfs/
```
Comment (StefanJum) [inline elfloader-net/README.md:51]: Should we add the network interface setup here as well? The `ip link add...` commands?

**Review by StefanJum — COMMENTED** [review]
Works fine now, see the one comment.

File: elfloader-net/README.md (hunk 3 above)
Comment (razvand) [reply elfloader-net/README.md:56]: > Should we add the network interface setup here as well? The `ip link add...` commands?

solved, thanks

**Review by StefanJum — APPROVED** [review]
All good.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Pull request merged**